### PR TITLE
Added authorization with pundit

### PR DESCRIPTION
**Pull request by jeeheezy**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,12 +1,24 @@
 class ApplicationController < ActionController::Base
+  include Pundit
   before_action :authenticate_user!
 
   before_action :configure_permitted_parameters, if: :devise_controller?
 
+  # after_action :verify_authorized, unless: :devise_controller?
+  # after_action :verify_policy_scoped, only: :index, unless: :devise_controller?
+
```
Comment (thierrychau) [inline app/controllers/application_controller.rb]: You do want these not commented out. It's our safeguard to make sure we have policies set up

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -23,6 +24,13 @@ def edit
 
   # POST /comments or /comments.json
   def create
+    @photo = Photo.find(params.fetch(:comment).fetch(:photo_id))
+    unless CommentPolicy.new(current_user, @photo).create?
+      raise Pundit::NotAuthorizedError, "not allowed to create? this #{@photo.inspect}"
+    end
```
Comment (thierrychau) [inline app/controllers/comments_controller.rb:34]: Very nice!

File: app/controllers/photos_controller.rb [hunk 3]
```diff
@@ -62,16 +64,18 @@ def destroy
     # Use callbacks to share common setup or constraints between actions.
     def set_photo
       @photo = Photo.find(params[:id])
+    rescue ActiveRecord::RecordNotFound => _
+      redirect_to root_url
```
Comment (thierrychau) [inline app/controllers/photos_controller.rb:68]: Nice!

File: app/views/comments/_comment.html.erb [hunk 4]
```diff
@@ -10,11 +10,15 @@
       <p><%= comment.body %></p>
     </div>
     <div>
-      <% if current_user == comment.author %>
+
+      <%# if current_user == comment.author %>
+      <% if policy(comment).edit? %>
```
Comment (thierrychau) [inline app/views/comments/_comment.html.erb:15]: nice!

File: app/views/comments/_comment.html.erb [hunk 5]
```diff
@@ -10,11 +10,15 @@
...
         <%= link_to edit_comment_path(comment), class: "btn btn-link btn-sm text-muted" do %>
           <i class="fas fa-edit fa-fw"></i>
         <% end %>
+      <% end %>
 
+      <% if policy(comment).destroy? %>
```
Comment (thierrychau) [inline app/views/comments/_comment.html.erb:21]: nice!

File: app/views/users/show.html.erb [hunk 6]
```diff
@@ -4,7 +4,7 @@
   </div>
 </div>
 
-<% if !@user.private? || current_user.leaders.include?(@user) || current_user == @user %>
+<% if policy(@user).show_content? %>
```
Comment (thierrychau) [inline app/views/users/show.html.erb:7]: nice!

**Review by thierrychau — CHANGES_REQUESTED** [review]
Looks great overall! Just add back the `verify_authorized` and `verify_policy_scoped` in the application_controller

**Review by thierrychau — APPROVED** [review]

**Review by thierrychau — APPROVED** [review]
Great job!

File: app/controllers/comments_controller.rb (hunk 2 above)
Comment (wahnahbe) [reply app/controllers/comments_controller.rb:34]: interesting way to go about this

File: app/controllers/comments_controller.rb [hunk 7]
```diff
@@ -1,7 +1,12 @@
 class CommentsController < ApplicationController
   before_action :set_comment, only: %i[ show edit update destroy ]
-  before_action :ensure_current_user_is_owner, only: [:destroy, :update, :edit]
-  before_action :is_an_authorized_user, only: [:create]
+  # before_action :ensure_current_user_is_owner, only: [:destroy, :update, :edit]
+  # before_action :is_an_authorized_user, only: [:create]
+  before_action(except: [:new, :create]) { authorize(@comment || Comment) }
+  # have to add in skip_after for verify_authorized since I"m not using authorize helper and instead using what helper does manually
+  # pundit is dumb
+  skip_after_action :verify_authorized, only: :create
```
Comment (wahnahbe) [inline app/controllers/comments_controller.rb:8]: nice way to go around the authorize

**Pull request merged**